lib/introspection.js

```diff
@@ -1,91 +1,235 @@
 'use strict';
 
 const common = require('metarhia-common');
-const { last } = common;
-
-const { ALL_TYPES } = require('./types');
-
-const FUNC_TERMS = [') {', ') => {', ') => ('];
-const NAMED_LINES = ['Example:', 'Returns:', 'Hint:', 'Result:'];
-
-const indexing = s => term => s.indexOf(term);
-
-const parseLines = (
-  // Parse signature lines
-  s, // string, signature lines
-  signature // record, { title, description, parameters, comments }
-  // Returns: array of string
-) => {
-  let lines = s.split('\n');
-  lines.pop();
-  signature.title = (lines.shift() || '').replace('//', '').trim();
-  lines = lines.map(
-    d => d.trim().replace(/^(.*) \/\//, '$1:').replace(',:', ':')
-  );
-  for (let line of lines) {
-    if (line.startsWith('//')) {
-      line = line.replace(/^\/\/ /, '').trim();
-      if (NAMED_LINES.find(s => line.startsWith(s))) {
-        const [name, comment] = common.section(line, ': ');
-        signature.comments.push({ name, comment });
-      } else if (signature.parameters.length === 0) {
-        if (signature.description.length > 0) {
-          signature.description += '\n';
-        }
-        signature.description += line;
+const ALL_TYPES = new Set([
+  'boolean', 'null', 'undefined', 'number', 'string', 'symbol',
+  'any', 'Primitive', 'Object', 'Date', 'Function',
+  'Map', 'WeakMap', 'Set', 'WeakSet', 'Array',
+  'Int8Array', 'Uint8Array', 'Uint8ClampedArray',
+  'Int16Array', 'Uint16Array',
+  'Int32Array', 'Uint32Array',
+  'Float32Array', 'Float64Array',
+  'Error', 'EvalError', 'RangeError', 'ReferenceError',
+  'SyntaxError', 'TypeError', 'URIError',
+]);
+
+const NAMED_LINES = [
+  'Returns:', // returned value
+  'Example:', // example of usage
+  'Result:', // result of usage, should be used with `Example`
+  'Throws:', // thrown errors
+  'Deprecated:', // reasons for deprecation or previous names
+];
+
+// Search `text` for `fn` occurrences
+//   fn - <Function> to be searched
+//   text - <string> to be searched in
+// Returns: <Array> lines of comments before function
+const parseComments = (fn, text) => {
+  const fnData = fn.toString();
+  const fnIndex = text.indexOf(fnData);
+  const lines = text.slice(0, fnIndex)
+    .split('\n')
+    .map(line => line.trim());
+  const comments = [];
+
+  for (let i = lines.length - 2; i >= 0; i--) {
+    const line = lines[i].trimLeft();
+    if (!line.startsWith('//')) break;
+    comments.unshift(line.slice(3));
+  }
+  return comments.map(line => line.trimRight()).filter(line => line.length);
```

Comment:
I'd prefer an explicit check for the `length` not being equal to 0:
```suggestion
  return comments.map(line => line.trimRight())
    .filter(line => line.length !== 0);
```